### ISSUE TITLE
IMGuiHelper uses a destroyed renderable instance

### ISSUE THREAD

**Issue by baktery** [google/filament#718]
**Describe the bug**
ImGuiHelper destroys a renderable instance and later continues using it.

**To Reproduce**
It has been hard to find the issue, visual glitches or several dynamic vertex buffers when rendering was produced (both imgui and non-imgui vbs).

Easier way I found to reproduce the issue is to dynamically create an indexbuffer while some imgui screens are visible.

A minimal case to reproduce it on my computer (Windows 10, OpenGL driver, nVidia) is adding the following preRender function to the material_sandbox sample. Do not forget to register the function to the FilamentApp preRender callback:

```
static void preRender(Engine* engine, View*, Scene* scene,  filament::Renderer*) {
	std::list<IndexBuffer*> buffers;
	
	for (int i = 0; i < 10; ++i)
	{
		//Fence::waitAndDestroy(engine->createFence());
		IndexBuffer* indexBuffer = IndexBuffer::Builder()
			.indexCount(6)
			.build(*engine);

		buffers.push_back(indexBuffer);

		const static uint32_t indices[] = {
			0, 1, 2, 2, 3, 0
		};
		indexBuffer->setBuffer(*engine, IndexBuffer::BufferDescriptor(
			indices, indexBuffer->getIndexCount() * sizeof(uint32_t)));
	}

	for (IndexBuffer* buffer : buffers)
	{
		engine->destroy(buffer);
	}
}
```
**How I fixed it**
Inside
```
void ImGuiHelper::renderDrawData(ImDrawData* imguiData)
```
The following line must be removed:

```
    rcm.destroy(mRenderable);
```

**Expected behavior**
No visual glitches. Depending on the position of the planets, sometimes a OpenGL error was produced (filling the index buffer).

**Desktop (please complete the following information):**
 - OS: Windows 10
 - GPU: GTX 1070
 - Backend: OpenGL

**Additional context**

A unrelated question:

Why ImGuiHelper needs to sync when recreating the vertex/index buffers?

```
void ImGuiHelper::syncThreads() {
#if UTILS_HAS_THREADING
    if (!mHasSynced) {
        // This is called only when ImGui needs to grow a vertex buffer, which occurs a few times
        // after launching and rarely (if ever) after that.
        Fence::waitAndDestroy(mEngine->createFence());
        mHasSynced = true;
    }
#endif
}
```

**Comment by prideout** [comment]
Regarding syncThreads: Filament's renderer runs in its own thread, so all clients (both imgui and non-imgui) need to wait until the Filament renderer is done before destroying and recreating vertex buffers and index buffers that are used by active renderables.

Regarding ImGuiHelper: do you see visual glitches only after adding your special preRender function, or do you see visual glitches with our unaltered sample code?

**Comment by baktery** [comment]
Thanks for the syncthreads explanation.

Unaltered sample code works fine. No problems with any of the existing filament samples.

Glitches start appearing if imgui is drawing something and new index/vertex buffers are created (outside ImGuiBuilder) after the imgui renderable instance destruction has been done.

**Comment by prideout** [comment]
In your minimal provoking case, I think you need to add `Fence::waitAndDestroy(engine->createFence())` between the two loops, otherwise Filament's rendering thread may try to upload data to the GPU that has already been freed.

**Comment by baktery** [comment]
This alternative case is also a good example to produce the OpenGL error:

```
static void preRender(Engine* engine, View*, Scene* scene,  filament::Renderer*) {

	static bool s_done = false;

	if (!s_done)
	{
		s_done = true;

		IndexBuffer* indexBuffer = IndexBuffer::Builder()
			.indexCount(6)
			.build(*engine);
	}

}
```

If I understand properly how filament works, the destruction of the ImGuiHelper renderable entity is a bug. Check this part of ImGuiHelper::renderDrawData(ImDrawData* imguiData)

```
    // Recreate the Renderable component and point it to the vertex buffers.
    rcm.destroy(mRenderable);
    int bufferIndex = 0;
    int primIndex = 0;
    for (int cmdListIndex = 0; cmdListIndex < imguiData->CmdListsCount; cmdListIndex++) {
        const ImDrawList* cmds = imguiData->CmdLists[cmdListIndex];
        size_t indexOffset = 0;
        populateVertexData(bufferIndex,
                cmds->VtxBuffer.Size * sizeof(ImDrawVert), cmds->VtxBuffer.Data,
                cmds->IdxBuffer.Size * sizeof(ImDrawIdx), cmds->IdxBuffer.Data);
        for (const auto& pcmd : cmds->CmdBuffer) {
            const size_t capacity = mIndexBuffers[bufferIndex]->getIndexCount();
            const size_t required = indexOffset + pcmd.ElemCount;
            assert(required <= capacity);
            if (pcmd.UserCallback) {
                pcmd.UserCallback(cmds, &pcmd);
            } else {
                uint64_t skey = makeScissorKey(fbheight, pcmd.ClipRect);
                auto miter = scissorRects.find(skey);
                assert(miter != scissorRects.end());
                rbuilder
                        .geometry(primIndex, RenderableManager::PrimitiveType::TRIANGLES,
                                mVertexBuffers[bufferIndex], mIndexBuffers[bufferIndex],
                                indexOffset, pcmd.ElemCount)
                        .blendOrder(primIndex, primIndex)
                        .material(primIndex, miter->second);
                primIndex++;
            }
            indexOffset += pcmd.ElemCount;
        }
        bufferIndex++;
    }
    if (imguiData->CmdListsCount > 0) {
        rbuilder.build(*mEngine, mRenderable);
    }
```

mRenderable is destroyed and later is used here:

```
rbuilder.build(*mEngine, mRenderable);
```

I think that a proper behavior must be one of the following options:

- Or the entity is not destroyed and it is rebuilt (just comment the line as I suggested in the first message)

- Or the entity is destroyed and another entity is created, something like doing this:

```
  rcm.destroy(mRenderable);
  mRenderable = em.create();

```
instead of just

```
  rcm.destroy(mRenderable);
```

**Comment by prideout** [comment]
Keep in mind that Filament is an ECS.  This might clear things up for you:
- `rcm.destroy` destroys the renderable component of the entity; it does not destroy the entity.
- `rbuilder.build` *creates* the renderable component of the entity; it does not "use" the renderable component of the entity.

**Comment by baktery** [comment]
So, to know what I'm doing bad on my side, do you see anything incorrect in the following example?

```
static void preRender(Engine* engine, View*, Scene* scene,  filament::Renderer*) {

	static bool s_done = false;

	if (!s_done)
	{
		s_done = true;

		IndexBuffer* indexBuffer = IndexBuffer::Builder()
			.indexCount(6)
			.build(*engine);
	}

}
```

**Comment by baktery** [comment]
It produces the following OpenGL error if imgui is drawing something:

OpenGL error 501 (GL_INVALID_VALUE) in "void filament::OpenGLDriver::updateIndexBuffer(Driver::IndexBufferHandle, filament::Driver::BufferDescriptor &&, uint32_t, uint32_t)" at line 1636

**Comment by prideout** [comment]
Are you waiting on a fence before destroying the index buffer?

**Comment by baktery** [comment]
in the last example no index buffer destruction is done

**Comment by baktery** [comment]
Independent of calling setBuffer the OpenGL error always appear:

This one:

```
static void preRender(Engine* engine, View*, Scene* scene,  filament::Renderer*) {

	static bool s_done = false;

	if (!s_done)
	{
		s_done = true;

		IndexBuffer* indexBuffer = IndexBuffer::Builder()
			.indexCount(6)
			.build(*engine);

		const static uint32_t indices[] = {
			0, 1, 2, 2, 3, 0
		};
		indexBuffer->setBuffer(*engine, IndexBuffer::BufferDescriptor(
			indices, indexBuffer->getIndexCount() * sizeof(uint32_t)));
	}

}
```
and this one:

```
static void preRender(Engine* engine, View*, Scene* scene,  filament::Renderer*) {

	static bool s_done = false;

	if (!s_done)
	{
		s_done = true;

		IndexBuffer* indexBuffer = IndexBuffer::Builder()
			.indexCount(6)
			.build(*engine);

	}

}
```

Same error

**Comment by prideout** [comment]
I was not able to repro this after hacking on suzanne.cpp as seen here:

https://github.com/google/filament/compare/pr/hack_suzanne?expand=1

@baktery do you have a fork that we can look at for a repro?

**Comment by baktery** [comment]
I tried your suzanne code. ImGui is not rendering anything in that sample (or at least I don't know if there's any shortcut), it works fine on my side, no way to reproduce the issue with it.

Could you try with an example where ImGui renders some stuff by default? I attached my local material_sandbox.cpp file. Let me know if you can reproduce the issue on your side with it.

[material_sandbox.zip](https://github.com/google/filament/files/2770017/material_sandbox.zip)

**Comment by MSTF** [comment]
Same problem. When changing widgets in loop, I get weird lines on mac + metal.
On Android (opengl), it's works only first frame and then UI completely disappears.

I think I found why official samples works. In official samples;

=======
DRAW UI
SDL_Delay(16);
PRE RENDER
RENDER
POST RENDER
=======

It's looks weird to have delay between UI and main view... After adding similar delay in my code, it's works on mac+metal (no glitchy lines) but android. I'm still having problem with Android as UI gets disappeared and not rendering at all after first frame.

**Comment by MSTF** [comment]
Also forcing to re-create ui vertex buffers each loop fixes the issue on both Metal + OpenGL, it's looks like the issue related to what @prideout wrote earlier.

**Comment by prideout** [comment]
@MSTF, can you file a new issue with repro steps?
